### PR TITLE
Add gapil/compiler

### PR DESCRIPTION
**Pull request by ben-clayton**
Implements a true compiler for gapil .api files.

### CODE REVIEW

File: gapil/compiler/map.go [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+// Once the map hits > 80% capacity, we should rehash the map larger.
+//  Otherwise collisions will turn this into a linear search.
+
+// Insertion into the map is similar (assuming it doesn't exist)
+//   1) h = Hashed Key % capacity
+//   2) rehash if necessary
+//   3) find first bucket >= h, where used != full (mod capacity)
+//   4) Insert there, mark used = full
+
+// TODO: Investigate alternative rehash rule.
```
Comment (AWoloszyn) [inline gapil/compiler/map.go]: This is done now

Comment (AWoloszyn) [reply gapil/compiler/map.go]: Not the next todo though

File: gapil/compiler/program.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	Line   int
+	Column int
+}
+
+func (l Location) String() string {
+	return fmt.Sprintf("%v:%v:%v", l.File, l.Line, l.Column)
+}
+
+func (p *Program) Dump() string {
+	//	types := make([]string, 0, len(c.types))
```
Comment (AWoloszyn) [inline gapil/compiler/program.go]: Do we want to keep these comments around?

File: gapil/compiler/map.go (hunk 1 above)
Comment (ben-clayton) [reply gapil/compiler/map.go]: Removed.

File: gapil/compiler/program.go (hunk 2 above)
Comment (ben-clayton) [reply gapil/compiler/program.go]: Nope, not even sure what they were. Removed.

**Pull request merged**